client/extensions/woocommerce/app/settings/payments/payment-method-stripe.js

```diff
@@ -217,6 +394,147 @@ class PaymentMethodStripe extends Component {
+		// interested in trying connect flow, so give the user a means to
+		// switch to connect flow
+		if ( ! this.hasKeys() && this.state.hadKeysAtStart ) {
+			showConnectFlowLink = true;
+		}
+
+		// Sanity check - if we are already in key flow, don't show the key flow link
+		if ( this.state.userRequestedKeyFlow ) {
+			showKeyFlowLink = false;
+		}
+
+		// Sanity check - if we are already in connect flow, don't show the connect flow link
+		if ( this.state.userRequestedConnectFlow ) {
+			showConnectFlowLink = false;
+		}
+
+		// Sanity check - if we already are connected, don't show the connect flow link
+		if ( connectedUserID ) {
+			showConnectFlowLink = false;
+		}
+
+		// OK, make it so
+		if ( showKeyFlowLink ) {
+			buttons.push( {
+				action: 'switch',
+				label: <span>{ translate( 'I want to enter my own keys' ) }</span>,
+				onClick: this.onUserRequestsKeyFlow,
+				additionalClassNames: 'payments__method-stripe-force-flow is-borderless'
+			} );
+		} else if ( showConnectFlowLink ) {
+			buttons.push( {
+				action: 'switch',
+				label: <span>{ translate( 'I want to use Stripe Connect instead' ) }</span>,
+				onClick: this.onUserRequestsConnectFlow,
+				additionalClassNames: 'payments__method-stripe-force-flow is-borderless'
+			} );
+		}
+
+		// We always give the user a Cancel button
+		buttons.push( { action: 'cancel', label: translate( 'Cancel' ), onClick: onCancel } );
+
+		// Figure out if we should show a Connect button or a Done button
+		let isConnectButton = false;
+		let isDisabled = false;
+
+		// No account?
+		if ( ! connectedUserID ) {
+			// No keys at start, no keys now and user didn't force key flow?
+			if ( ! this.state.hadKeysAtStart && ! this.hasKeys() && ! this.state.userRequestedKeyFlow ) {
+				// Give them a Connect button
+				isConnectButton = true;
+				isDisabled = true; // TODO - enable this in a subsequent PR to allow connection
+			} else {
+				// Give them a Done button, but enabled based on validity of key fields
+				isDisabled = ! hasStripeValidCredentials( method );
+			}
+		}
+
+		// Did the user request connect flow explicitly? Give them a Connect button
+		if ( this.state.userRequestedConnectFlow ) {
+			isConnectButton = true;
+			isDisabled = true; // TODO - enable this in a subsequent PR to allow connection
+		}
+
+		buttons.push( {
+			action: isConnectButton ? 'connect' : 'save',
+			disabled: isDisabled,
+			label: isConnectButton ? translate( 'Connect' ) : translate( 'Done' ),
+			onClick: isConnectButton ? this.onConnect : this.onDone,
+			isPrimary: true
+		} );
+
+		return buttons;
+	}
+
+	////////////////////////////////////////////////////////////////////////////
+	// And render brings it all together
+
+	render() {
+		return (
+			<Dialog
+				additionalClassNames="payments__dialog woocommerce"
+				buttons={ this.getButtons() }
+				isVisible>
+				{ this.renderHeading() }
+				{ this.possiblyRenderSetupPrompt() }
```

Comment:
I'm not sure how I feel about all the `possiblyXXX` functions, but I don't immediately have a better idea. It's just that there's really no way to know what this render function is actually rendering without tracing back a lot of code.

Reply:
Its a pattern we used a lot on /me and on hydra work. Its meant to 1) keep the render method from getting so large as to be difficult to restructure and 2) convey that the call might return markup or might return nothing.